### PR TITLE
[OM-88273]: Initial version of GitOpsMode CRD

### PR DESCRIPTION
**Pull request by amd-ibm**
## Intent

Introduce new CRD to allow override of default GitOps configuration.

## Background

Allow turbo users who deploy k8s applications using a GitOps pipeline (ArgoCD) to override Kubeturbo's default GitOps configuration, such that they can specify the commit mode, etc when executing actions. 

Most of the code container here is boilerplate generated by Kubebuilder. Here are the files that you should focus on:
* `api/v1alpha1/gitops_types.go`
* `config/samples/gitops_v1alpha1_gitops.yaml`

## Testing

Deployed CRD various use cases with variations of the sample CR. This includes
* providing a valid spec with and without the optional `credentials` property
* providing an invalid commit mode type (kubectl command fails with error such as `The GitOps "gitops-sample" is invalid: spec.config.commitMode: Unsupported value: "pr": supported values: "direct", "request"`)
* ~~providing an empty `apps` list (kubectl command fails with error such as `The GitOps "gitops-sample" is invalid: spec.config.apps: Invalid value: 0: spec.config.apps in body should have at least 1 items`~~
* failure to include all required fields for `credentials` property (kubectl command fails with error such as `error: error validating "gitops_v1alpha1_gitops.yaml": error validating data: [ValidationError(GitOps.spec.config[0].credentials): missing required field "secretName" in io.turbonomic.gitops.v1alpha1.GitOps.spec.config.credentials, ValidationError(GitOps.spec.config[0].credentials): missing required field "secretNamespace" in io.turbonomic.gitops.v1alpha1.GitOps.spec.config.credentials, ValidationError(GitOps.spec.config[0].credentials): missing required field "username" in io.turbonomic.gitops.v1alpha1.GitOps.spec.config.credentials]; if you choose to ignore these errors, turn validation off with --validate=false`)

### CODE REVIEW

**Comment by ading1977** [comment]
I think the Background section is out-dated, right?

**Comment by amd-ibm** [comment]
> I think the Background section is out-dated, right?

Yeah. I copied from the previous PR, but forgot to update. I'll get that updated.

**Review by ading1977 — APPROVED** [review]

File: api/v1alpha1/gitops_types.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+// GitOpsSpec defines the desired state of GitOps configuration
+type GitOpsSpec struct {
+	// Overrides the default GitOps configuration with custom configuration for the specified app(s).
+	Configutaion []Configuration `json:"config"`
+}
+
+// CommitMode describes how Turbonomic events will be processed.
+// +kubebuilder:validation:Enum=direct;request
+type CommitMode string
```
Comment (pankajbatra1) [inline api/v1alpha1/gitops_types.go:36]: Please elaborate on the testing section.
Specifically, please list both positive and negative tests that you did for CRD and CR creation.

File: PROJECT [hunk 2]
```diff
@@ -0,0 +1,15 @@
+domain: turbonomic.io
```
Comment (pankajbatra1) [inline PROJECT:1]: In reviews like this, please list the files that need to be reviewed carefully in the description section

File: api/v1alpha1/gitops_types.go (hunk 1 above)
Comment (amd-ibm) [reply api/v1alpha1/gitops_types.go:36]: Added additional information about some of the various test cases.

File: PROJECT (hunk 2 above)
Comment (amd-ibm) [reply PROJECT:1]: Updated the description to highlight the non-boilerplate files to focus on.

File: api/v1alpha1/gitops_types.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+type Configuration struct {
+	// Specifies the GitOps commit mode.
+	// Valid values are:
+	// - "direct": actions will produce commit directly within the underlying repository without creating a pull/merge request;
+	// - "request": actions will result in a pull/merge request being creating within the underlying repository
+	// +kubebuilder:default:=request
+	CommitMode CommitMode `json:"commitMode,omitempty"`
+	// Specifies the credentials for the underlying repository (CURRENTLY UNSUPPORTED)
+	// +optional
```
Comment (pankajbatra1) [inline api/v1alpha1/gitops_types.go:53]: I would have thought that Credentials is not optional and apps is optional.

**Pull request merged**